### PR TITLE
Add global settings for VM migration flags

### PR DESCRIPTION
**Pull request by ravening**
### Description
Provide global settings or vm migration flags for kvm
so that admins can configure the values

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### Screenshots (if appropriate):

Screenshot 2021-02-03 at 11 38 23

### CODE REVIEW

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -422,6 +426,13 @@
                         "Indicates whether the host in down state can be put into maintenance state so thats its not enabled after it comes back.",
                         true, ConfigKey.Scope.Zone, null);
 
+    protected final ConfigKey<Integer> KvmVmMigrateSpeed = new ConfigKey<>("Advanced", Integer.class, KVM_VM_MIGRATE_SPEED, "-1",
+            "set the vm migrate speed (in MiB/s) on KVM. By default, it will try to guess the speed of the guest network (in MBps).", true);
+    protected final ConfigKey<Integer> KvmVmMigrateDowntime = new ConfigKey<>("Advanced", Integer.class, KVM_VM_MIGRATE_DOWNTIME, "-1",
+            "Sets maximum tolerable time in milliseconds for which the domain is allowed to be paused at the end of live migration on KVM.", true);
+    protected final ConfigKey<Integer> KvmVmMigratePauseAfter = new ConfigKey<>("Advanced", Integer.class, KVM_VM_MIGRATE_PAUSE_AFTER, "-1",
+            "Set an upper limit in milliseconds for how long live migration on KVM should wait, at which point VM is paused and migration will finish quickly.  Less than 1 means disabled.", true);
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: these belong in the hypervisor plugin for KVM, not here.

**Comment by sureshanaparti** [comment]
@ravening any update on this PR?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 1 above)
Comment (ravening) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: @DaanHoogland do you know the right place where can I add it? I didnt find any file for kvm specific global settings.

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: I'd say in the libvirt guru or - resource, I'd say. But to tell you the truth, I really don't know. As they are used in the AgentManager, maybe there. But for sure these shouldn't be initialised in `ConfigurationManagerImpl.initMessageBusListener()`.
For vmware there are such `ConfigKey`'s stored in the `VmwareGuru` and in a `VmwareManager`. Both seem resonable. In the KVM case the `KVMGuru` (which actually should be in the kvm hypervisor plugin, soit).

Comment (weizhouapache) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: > I'd say in the libvirt guru or - resource, I'd say. But to tell you the truth, I really don't know. As they are used in the AgentManager, maybe there. But for sure these shouldn't be initialised in `ConfigurationManagerImpl.initMessageBusListener()`.
> For vmware there are such `ConfigKey`'s stored in the `VmwareGuru` and in a `VmwareManager`. Both seem resonable. In the KVM case the `KVMGuru` (which actually should be in the kvm hypervisor plugin, soit).

agree with @DaanHoogland 
`KVMGuru` looks like better place.

**Comment by rhtyd** [comment]
Ping @ravening can you address conflicts?

**Comment by sureshanaparti** [comment]
Hi @ravening can you fix the conflicts, thanks!

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 1 above)
Comment (ravening) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: @DaanHoogland @weizhouapache `KVmguru` doesnt have any code related to configKeys. shall i add that also?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:435]: Yes, makes sense to me @ravening

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 2]
```diff
@@ -654,6 +658,54 @@ public String interpret(final BufferedReader reader) throws IOException {
...
+    /**
+     * setLiveMigrateFlags
+     *
+     * Sets the user configured live migration flags
+     *
+     * @param params
+     */
+    public void setLiveMigrateFlags(final Map<String, String> params) {
+        if (params.get(KVM_VM_MIGRATE_SPEED) != null) {
+            _migrateSpeed = NumbersUtil.parseInt(params.get(KVM_VM_MIGRATE_SPEED), -1);
```
Comment (nvazquez) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:670]: What about `NumbersUtil.parseInt(params.get(KVM_VM_MIGRATE_SPEED), getDefaultMigrateSpeed());` instead? The next `if` block won't be needed

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -453,6 +457,12 @@
             "Max length of vm userdata after base64 decoding. Default is 32768 and maximum is 1048576", true);
     public static final ConfigKey<Boolean> MIGRATE_VM_ACROSS_CLUSTERS = new ConfigKey<Boolean>(Boolean.class, "migrate.vm.across.clusters", "Advanced", "false",
             "Indicates whether the VM can be migrated to different cluster if no host is found in same cluster",true, ConfigKey.Scope.Zone, null);
+    protected final ConfigKey<Integer> KvmVmMigrateSpeed = new ConfigKey<>("Advanced", Integer.class, KVM_VM_MIGRATE_SPEED, "-1",
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:460]: Would it make sense to reduce the scope of these setting to Cluster level?

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
Hi @ravening, Can you address the outstanding comments. Thanks.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 1 above)
Comment (ravening) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: @DaanHoogland @weizhouapache moved the settings to `KVMguru`

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java (hunk 2 above)
Comment (ravening) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: changed

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 3 above)
Comment (ravening) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: changed to cluster scope

**Comment by ravening** [comment]
> Hi @ravening, Can you address the outstanding comments. Thanks.

@sureshanaparti done

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by davidjumani — APPROVED** [review]

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 4]
```diff
@@ -1759,8 +1764,11 @@ public void processConnect(final Host host, final StartupCommand cmd, final bool
             if (cmd instanceof StartupRoutingCommand) {
                 if (((StartupRoutingCommand)cmd).getHypervisorType() == HypervisorType.KVM || ((StartupRoutingCommand)cmd).getHypervisorType() == HypervisorType.LXC) {
                     Map<String, String> params = new HashMap<String, String>();
-                    params.put(Config.RouterAggregationCommandEachTimeout.toString(), _configDao.getValue(Config.RouterAggregationCommandEachTimeout.toString()));
                     params.put(Config.MigrateWait.toString(), _configDao.getValue(Config.MigrateWait.toString()));
+                    params.put(ROUTER_AGGREGATION_COMMAND_EACH_TIMEOUT, _configDao.getValue(ROUTER_AGGREGATION_COMMAND_EACH_TIMEOUT));
+                    params.put(KVM_VM_MIGRATE_SPEED_STRING, _configDao.getValue(KVM_VM_MIGRATE_SPEED_STRING));
```
Comment (weizhouapache) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:1769]: KVM_VM_MIGRATE_SPEED is a cluster-level setting, but global config is uses here.

**Comment by davidjumani** [comment]
@ravening Can you have a look at the test failures? Thanks

**Comment by ravening** [comment]
> @ravening Can you have a look at the test failures? Thanks

@davidjumani one of the test is failing because of `Exception during migrate: org.libvirt.LibvirtException: numerical overflow: bandwidth must be less than 8796093022208` which is totally not related to my change.

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Hi @ravening can you please fix the conflicts?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Pull request closed without merging**